src/ws/session.c

```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
 
   for (i = 0; i < num_msg; i++)
     {
-      if (msg[i]->msg_style == PAM_PROMPT_ECHO_OFF)
+      if (msg[i]->msg_style == PAM_PROMPT_ECHO_OFF &&
+          *password != NULL)
+        {
+            debug ("answered pam password prompt");
+            resp[i].resp = *password;
+            resp[i].resp_retcode = 0;
+            *password = NULL;
+        }
+      else if (msg[i]->msg_style == PAM_ERROR_MSG)
         {
-          if (*password == NULL)
+          if (asprintf (&err_msg, "%s ", msg[i]->msg) < 0)
```

Comment:
Why are we reallocating this? Would a const char suffice? If not, then a comment would be good.